hazelcast/src/main/java/com/hazelcast/internal/probing/ProbeRegistryImpl.java

```diff
@@ -0,0 +1,681 @@
+package com.hazelcast.internal.probing;
+
+import static com.hazelcast.internal.probing.CharSequenceUtils.appendEscaped;
+import static com.hazelcast.internal.probing.CharSequenceUtils.appendUnescaped;
+import static com.hazelcast.util.StringUtil.getterIntoProperty;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import com.hazelcast.internal.metrics.Probe;
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.internal.probing.ProbingCycle.Tagging;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.util.Clock;
+
+public final class ProbeRegistryImpl implements ProbeRegistry {
+
+    private static final ILogger LOGGER = Logger.getLogger(ProbeRegistryImpl.class);
+
+    private static final Object[] EMPTY_ARGS = new Object[0];
+
+    private static final Map<Class<?>, ProbeAnnotatedType> PROBE_METADATA = 
+            new ConcurrentHashMap<Class<?>, ProbeAnnotatedType>();
+
+    private final Deque<ProbeSourceEntry> sources = new ConcurrentLinkedDeque<ProbeSourceEntry>();
+
+    @Override
+    public void register(ProbeSource source) {
+        for (ProbeSourceEntry e : sources) {
+            if (e.source.equals(source)) {
+                LOGGER.info("Probe source tried to register more then once: "
+                        + source.getClass().getSimpleName());
+                return; // avoid adding the very same instance more then once
+            }
+        }
+        sources.add(new ProbeSourceEntry(source));
```

Comment:
There is a data race here.

Reply:
Do we have concurrent threads that initialize the cluster? I thought this is single threaded. Reason to use concurrent collection is that threads using the collection will be another one.

Reply:
Fixed.